storm-server/src/main/java/org/apache/storm/daemon/nimbus/TimedWritableByteChannel.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+package org.apache.storm.daemon.nimbus;
+
+import com.codahale.metrics.Timer;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+
+import org.apache.storm.metric.timed.TimedResource;
+
+public class TimedWritableByteChannel extends TimedResource<WritableByteChannel> implements WritableByteChannel {
+
+    public TimedWritableByteChannel(WritableByteChannel measured, Timer timer) {
+        super(measured, timer);
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+        return getMeasured().write(src);
+    }
+
+    @Override
+    public boolean isOpen() {
+        return getMeasured().isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            super.close();
+        } catch (Exception e) {
+            throw (IOException) e;
```

Comment:
If you know this is an IOException, you should change the caught type instead.

Reply:
The superclass `close` can throw any Exception (e.g. from `getMeasured().close()`), so I don't think you can do this cast safely. You probably need to wrap any non-IOException and rethrow as unchecked.

If you want to avoid this, you would probably need to ensure `Timed<T extends Closeable>` instead of `Autocloseable`, not sure how many classes only implement the latter.

Reply:
getMeasured() is guaranteed the return type of `WritableByteChannel` here I believe, because this class extends `TimedResource<WritableByteChannel>`.

Reply:
getMeasured().close() should be instantiated with `Channel#close()` here I believe, because the class extends `TimedResource<WritableByteChannel>`. `getMeasured` should always return type `WritableByteChannel`

Reply:
You're right. Nevermind then. I can't think of a better structure either.

Reply:
I'll comment this to improve it's readability